### ISSUE TITLE
Doc states that ESOCKETTIMEOUT should appear as ETIMEOUT but it doesn't

### ISSUE THREAD

**Issue by mcortesi** [request/request#2320]
Maybe this is a documentation issue or a code issue.

The documentation [here](https://github.com/request/request#timeouts) states that:

> These two situations have widely different implications for what went wrong with the request, so it's useful to be able to distinguish them. **You can detect timeout errors by checking err.code for  an 'ETIMEDOUT' value.** Further, you can detect whether the timeout was a connection timeout by checking if the err.connect property is set to true.

But [code](https://github.com/request/request/blob/290111d4e94ac9f6c8d3abecbc8920c96d1b32be/request.js#L776) says different, it sends `err.code` as `ESOCKETTIMEOUT`

**Comment by drmrbrewer** [comment]
It seems to me that the `timeout` value you pass in via the `options` object is used to set two different timeouts, with two different error codes:

    ETIMEDOUT: will throw if the server takes more than `timeout` to write the HTTP status and headers (corresponding to the on('response') event on the client).

    ESOCKETTIMEDOUT: the amount of time to wait *between* bytes sent from the server once connected.  In particular, it's useful for erroring if the server fails to send data halfway through streaming a response

I may be wrong, but I too am getting `ESOCKETTIMEDOUT` sometimes and `ETIMEDOUT` at other times... and the above is all I have found so far to try to explain it.  Maybe the docs just need to be updated?

**Comment by krukid** [comment]
Yep, spent an hour googling how to set read timeout on `request` and `request-promise`, it seems like the docs lied to me. Source sets the same timeout value for both connect and read, not **just** the connect.

**Comment by jme783** [comment]
+1 here, I too was confused my the documentation here. I think the commentary here is correct: The error code is in fact different if it is a connection timeout vs. a read timeout. @mikeal and/or @simov can you weigh in here? I am happy to make a P/R to update the "Timeout" section of the docs to make this more accurate.

**Comment by aryehb** [comment]
I know this library is deprecated, but I'd like to document my findings in case anyone else is dealing with legacy code that still uses this library.

>ETIMEDOUT will throw if the server takes more than timeout to write the HTTP status and headers (corresponding to the on('response') event on the client).

That is indeed what the comment says [here](https://github.com/request/request/blob/3c0cddc7c8eb60b470e9519da85896ed7ee0081e/request.js#L841-L844).

However, looking at that code, it looks like the timer is removed once the socket is connected and emits the `connect` event, which is even before the request was written to the socket.

Based on that, I would say that `ETIMEDOUT` will be thrown if it takes more than `timeout` from when the socket is created (corresponding to the `socket` event) until the connection is established (signaled by the `connect` event on the socket).

---

>ESOCKETTIMEDOUT is associated with the amount of time to wait between bytes sent from the server once connected. In particular, it's useful for erroring if the server fails to send data halfway through streaming a response

This is also from the [comments](https://github.com/request/request/blob/3c0cddc7c8eb60b470e9519da85896ed7ee0081e/request.js#L808-L812).

Looking at the code there, it looks like it calls Node's HTTP [`ClientRequest.setTimeout`](https://nodejs.org/api/http.html#http_request_settimeout_timeout_callback), which calls the socket's [`setTimeout`](https://nodejs.org/api/net.html#net_socket_settimeout_timeout_callback), which is documented that it 

>Sets the socket to timeout after timeout milliseconds of inactivity on the socket.

I presume this would also include the time from when the connection is established until the server starts writing to the socket.

---

tl;dr:

If I understand correctly, this is the difference between the error codes:

`ETIMEDOUT` will be thrown if it takes more than `timeout` milliseconds from when the socket is created until the connection is established.

`ESOCKETTIMEDOUT` will be thrown when there is `timeout` milliseconds of inactivity on the socket.
